### ISSUE TITLE
return connection along with value in JavaScript promises

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#4026]
The promises returned by the JavaScript driver currently returns only the value explicitly called for, i.e. `get()` returns the record specified. But if you are trying to include a `connect()` call in the promise chain this means you have to stash the connection in a global (or at least outer scoped) variable in order to use it in links other than the one immediately following the `connect()` one. For example:

```javascript
var globalConn = null;
r.connect().then(function(conn) {
    globalConn = conn; // required to use in subsequent links
    return r.get(1).run(conn);
}).then(function(value, conn) {
    // conn is undefined, so globalConn is is the only way of getting the connection
})
```

However, the BlueBird module that we use for Promises does allow for multiple values using the [`spread`](https://github.com/petkaantonov/bluebird/blob/master/API.md#option-spread) option. This would allow for promise chains with multiple ReQL queries in it without requiring a (semi) global variable.

**Comment by danielmewes** [comment]
Chaining the connection through sounds reasonable.
@deontologician what do you think?

**Comment by deontologician** [comment]
Yeah, I think it's a great idea. If we don't do #3771 we should do this

**Comment by neumino** [comment]
@larkost you can use `bind`.

```js
r.connect().bind({}).then(function(c) {
  this.c = c;
  return query.run(this.c);
.then(function(result) {
  this.c.close();
})
```

See https://github.com/petkaantonov/bluebird/blob/master/API.md#binddynamic-thisarg---promise

**Comment by larkost** [comment]
Thanks @neumino, I hadn't found that way yet. I still think that we should give the connection as a second returned item as that allows for the cleanest code.

**Comment by stuartpb** [comment]
It allows for the cleanest code, until you need *another* thing in your promise scope (such as the results from a previous operation), at which point you're in the same predicament. This is an inherent issue with the promise model: @neumino is right that Bluebird's `.bind()` is the best solution.

**Comment by danielmewes** [comment]
@stuartpb Thanks for your feedback. It seems to me like the connection is something that you have to bind / chain through especially frequently though, so it might still make sense to special case it to make that case more convenient.
